src/stan/io/json/json_data_handler.hpp

```diff
@@ -7,174 +7,527 @@
 #include <stan/io/var_context.hpp>
 #include <cctype>
 #include <iostream>
+#include <ostream>
 #include <limits>
 #include <map>
+#include <numeric>
 #include <sstream>
 #include <string>
 #include <utility>
 #include <vector>
+#include <boost/algorithm/string.hpp>
 
 namespace stan {
 
 namespace json {
 
-typedef std::map<std::string,
-                 std::pair<std::vector<double>, std::vector<size_t>>>
-    vars_map_r;
+typedef std::pair<std::vector<double>, std::vector<size_t>> var_r;
+typedef std::pair<std::vector<int>, std::vector<size_t>> var_i;
 
-typedef std::map<std::string, std::pair<std::vector<int>, std::vector<size_t>>>
-    vars_map_i;
+typedef std::map<std::string, var_r> vars_map_r;
+typedef std::map<std::string, var_i> vars_map_i;
+
+/** Enum of the kinds of structures the handler needs to manage.
+ *  Determined by the initial sequence of start elements following
+ *  the top-level set of keys in the JSON object.
+ */
+struct meta_type {
+  enum {
+    SCALAR = 0,           // no start elements
+    ARRAY = 1,            // one or more "["
+    TUPLE = 2,            // one or more "{"
+    ARRAY_OF_TUPLES = 3,  // one or more "[" followed by "{"
+  };
+};
+
+/** Enum which tracks handler events.
+    Used to identify first and last slot of a tuple.
+*/
+struct meta_event {
+  enum {
+    OBJ_OPEN = 0,   // {
+    OBJ_CLOSE = 1,  // }
+    KEY = 2,
+  };
+};
```

Comment:
It seems like we're only ever actually care if we're in OBJ_OPEN when we see a key, so do we need to store the additional states? I still have a hard time following the `save_key_value_pair` logic this is tied to